### ISSUE TITLE
Stochastic memory leak (?) in subtraction

### ISSUE THREAD

**Issue by bforsbe** [3dem/relion#673]
on a96ba95, subtraction jobs sometime take an enormous amount of memory, and sometimes very little. For instance, 

2 MPI subtract 512 box, rescale to 256, 50k particles: 8min on 1 node, using ~5GB

2 MPI subtract 512 box, rescale to 256, 500k particles: stalls after starting to swap on 2 or 4 nodes with 1 MPI/node, linear increase in mem use that within minutes swaps at >128GB on some nodes, and virtually no mem on others. 

2 MPI subtract 512 box, rescale to 256, 500k particles:  1 nodes with 1 MPI/node, no mem ramp and uses ~5GB.

Basically, mem use varies wildly and unpredictably. The only reason a single MPI would create enormous and ramping mem-use is for output alloction for HUGE output stacks, but mem use does not seem to correlate with particle numbers on a single node. My guess is a memory leak or stochastic overflow in a mem-allocation, or something similar.

**Comment by bforsbe** [comment]
Any job that runs for long enough eventually sees a point where the RAM ramps up to max within minutes. Short jobs finish before that happens. 
![Screenshot from 2020-09-15 09-42-59](https://user-images.githubusercontent.com/7627666/93180799-f726d280-f737-11ea-878e-73af8e8591e5.png)

**Comment by bforsbe** [comment]
For comparison, this is a job with fewer particles, but otherwise the same, that does not run into the issue:
![Screenshot from 2020-09-15 12-57-07](https://user-images.githubusercontent.com/7627666/93202558-2ac32600-f753-11ea-9576-f5b0c087af0d.png)

**Comment by biochem-fan** [comment]
Can you run valgrind to detect which variable is leaking?

**Comment by bforsbe** [comment]
I worked around this instability by splitting the file with particles to be subtracted into 20 pieces and ran 20 subtractions jobs on the same node at the same time, then joined them back together and ran a newstack_create. This peaked in mem-use near 120GB intially (during setup with FFTs), but stabilized on ~50GB i.e. around 2.5GB for each subtraction process. I mention this only to indicate that there is no reason why subtraction should demand as much memory as it does occasionally.

**Comment by bforsbe** [comment]
I tried running valgrind but I'm having issues with our cluster installation of it. Trying Allinea-DDT now. This is behaving really strange though... here's a recent run that stops short of swapping, but only just. Is relion_particle_subtract estimting the available space and making a mess of it sometimes or what? 
![Screenshot from 2020-09-29 21-03-58](https://user-images.githubusercontent.com/7627666/94610160-c4342100-0297-11eb-852a-99bf9a50965a.png)

**Comment by bforsbe** [comment]
It's also problematic that this is really stochastic and seems to be evident for larger inputs and MPI-runs, where debugging is notoriously more difficult.

**Comment by biochem-fan** [comment]
I don't think looking at the memory usage reported by `top` and `free` is the right way. `malloc` gets memory from the kernel by system calls and does not necessarily return it upon `free` so that the space can be re-used in the next `malloc`.

If there is really memory leak in `relion_particle_subtract`, it should happen regardless of the input size. I will run valgrind on our tutorial dataset.

**Comment by biochem-fan** [comment]
Valgrind found nothing suspicious.

Why you cannot run valgrind? It does not require root.

**Comment by bforsbe** [comment]
libc6-dbg is not intalled on our cluster apparently, and valgrind needs it. 

This is not via top or free, and it definitely swapped in the first few runs I reported. The nodes seized and became unresponsive and had to be rebooted.

**Comment by bforsbe** [comment]
The runs I'm doing use the  ```--data```, ```--recenter_on_mask``` and ```--new_box``` flags. Maybe if you use them through valgrind? 

But as you see in the first image I attached it really looks like a rare event...you might need a big input file to see it.

**Comment by bforsbe** [comment]
I'll try to get valgrind running on my data because allinea is taking WAY too long.

**Comment by biochem-fan** [comment]
> This is not via top or free

Isn't the source of your plots the same as `top` and `free` (i.e. same field in `/proc`)? It looks like node level information.

> The runs I'm doing use the --data, --recenter_on_mask and --new_box flags

I tried again with them but found nothing.

> it really looks like a rare event...

Then I will leave this to you.

**Comment by eriklindahl** [comment]
Björn: Instead of running in a debugger, you could try using a recent version of Clang and simply enable the leak sanitiser part of the address sanitiser when building. It will be slow, but nowhere near as slow as using a debugger.

1) Add the flag -fsanitize=address when building with clang.

2) Before running RELION, set the environment variable

ASAN_OPTIONS=detect_leaks=1



Alternatively, recent versions might also be able to use the leak sanitiser directly with -fsanitize=leak.

**Comment by bforsbe** [comment]
@biochem-fan What is a reasonable estimate / example of expected time for subtraction in you hands, given a) Box-size b) MPIs, and c) particle number? I'll try to match it for a test case and see if our cluster is way off or if it makes sense. 

For relion_refine I have solid intuitions for adequate performance and expected memory use but not for the subtractor.

**Comment by bforsbe** [comment]
AAAARGH.  

Same run. Same MPI-setting. Same input. Same hardware. Same debug-build binary. 

On left: Through allinea-ddt to catch and see what the mamory allocations are. 
On right: Run through relion-GUI job submission. 

![Screenshot from 2020-09-30 10-49-02](https://user-images.githubusercontent.com/7627666/94670922-15312d00-030b-11eb-8b49-d542d4b671eb.png)

**Comment by bforsbe** [comment]
I was able to attach to a running process like the one on the right, and while allinea doesn't permit me to see memory stacks of attached processes, I find by setting a breakpoint inside ```SubtractOneParticle()```  that the memory use reported by 
```pmap <PIDZ> ``` 
increases by 2048k bytes between breaks. Stepping through ``` SubtractOneParticle()```  and monitoring it, 3 allocations are made:

1. https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/particle_subtractor.cpp#L467   2048k
2. https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/particle_subtractor.cpp#L532   4112k
3. https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/particle_subtractor.cpp#L580   2056k

At the end of the function, 6168k are released. This would indicate that the first allocation does NOT get freed.

**Comment by bforsbe** [comment]
The stack trace for the "leaky" allocation is 
```
#7 main (argc=11, argv=0x7ffe3a5e54e8) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/apps/particle_subtract_mpi.cpp:47 (at 0x0000000000514a74)
#6 ParticleSubtractor::run (this=0x7ffe3a5e31b0) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/particle_subtractor.cpp:324 (at 0x00000000005654fc)
#5 ParticleSubtractor::subtractOneParticle (this=0x7ffe3a5e31b0, part_id=11049, imgno=0, counter=5526) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/particle_subtractor.cpp:467 (at 0x0000000000566b31)
#4 Image<double>::read (this=0x7ffe3a5e2800, name=..., readdata=true, select_img=-1, mapData=false, is_2D=false) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/image.h:510 (at 0x000000000056bcd3)
#3 Image<double>::_read (this=0x7ffe3a5e2800, name=..., hFile=..., readdata=true, select_img=80, mapData=false, is_2D=false) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/image.h:1390 (at 0x000000000056fcde)
#2 Image<double>::readMRC (this=0x7ffe3a5e2800, img_select=80, isStack=true, name=...) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/rwMRC.h:271 (at 0x0000000000575a3e)
#1 Image<double>::readData (this=0x7ffe3a5e2800, fimg=0x1964cf60, select_img=80, datatype=Float, pad=0) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/image.h:993 (at 0x000000000057dd8e)
#0 MultidimArray<double>::coreAllocateReuse (this=0x7ffe3a5e2800) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/multidim_array.h:741 (at 0x000000000057e6cd)
```
pointing to RELION_ALIGNED_MALLOC: 
https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/multidim_array.h#L62

Perhaps RELION_ALIGNED_FREE is not called where it is supposed to be?

**Comment by bforsbe** [comment]
Reverting src/multidimarray.h to 6782e81bd1 seems to get rid of the apparent memory leak. I would conclude that the offending commit is a67564a7, where the method of allocation/deallocation of multidimarrays was changed. I cannot possibly say why this leak is seemingly random, but that's seems to be where it is coming from. Perhaps @biochem-fan can advise what the change was more in-depth, or @eriklindahl can offer some experience or insight as for the difference between the allocation methods and how it might lead to occasional leaks?

**Comment by eriklindahl** [comment]
a67564a seems to do several major NO-NOs.

1) One cannot just replace a call to new with a c-style memory allocation and then casting the pointer. The new() call will construct an object, while the latter just takes the memory and the cast says "this is an object of the right type". No constructors will be called.

2) Same for delete - that involves calling the destructor, which is where magic - such as deallocating memory - often happens.

3) C++ needs to separate delete from delete[], where the latter operates on an array of elements and destructs+frees them all. Merely calling free() on the memory will not accomplish that.

**Comment by biochem-fan** [comment]
But `data` is just a pointer to a basic type (e.g. `float` or `double`), not objects. We don't have constructors or destructors. Of course we cannot `new` and then `free` or `malloc` then `delete` but here the usage is consistent.

**Comment by bforsbe** [comment]
Well in my hands the difference in both speed and memory is quite astonishing, in favor of the "old" multidimarray allocation method. I can run 8 MPIs per node (2 nodes) and and be done with 300k particles subtractions in ~15min (using a total of ~17GB RAM), instead of capping out on RAM after 10 minutes and not getting a time estimate for 20hours.

**Comment by biochem-fan** [comment]
By the way, what compiler are you using?

**Comment by bforsbe** [comment]
gcc 7.3

**Comment by bforsbe** [comment]
If I had to guess (and it's really a guess), perhaps the old method is subject to destruction by going out of scope or by default mechanism that the newer is not. If coreDeallocate is called when needed the new method presumably works fine. There seems to be cases where it does not and it might be down to bad initialization of objects or members that control default or fallback deallocation behaviour? That would be one way of reconciling the "stochastic" nature of it. 

Would it be reasonable/informative to track where/how it gets deallocated in the old version?

**Comment by biochem-fan** [comment]
The destructor of `Image` calls `MultidimArray`'s `clear()`, which then calls `coreDeallocate()`, which performs `RELION_ALIGNED_FREE(data)`.

One concern is when `mmapOn` is true.  This is a legacy from Xmipp (should be removed) and I think this is never used in RELION but can you confirm this is always false?

**Comment by bforsbe** [comment]
It was false during allocation in the debug-case where I saw that it failed to deallocate. I can try to verify this is also the case under deallocation, but it should be false then too.

**Comment by bforsbe** [comment]
data appear to be NULL here
https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/multidim_array.h#L763

so coreDeallocate() just skips the call to RELION_ALIGNED_FREE(data).

**Comment by bforsbe** [comment]
That is, mmapOn is false, but it never makes it that far.

**Comment by bforsbe** [comment]
The destructor on the img multidimarray seems to be called at subtractOneParticle() exit, but it looks to have been resized to nothing by then. That's probably all the info I can supply at the moment. Let me know if there's anything else I can do.

**Comment by biochem-fan** [comment]
Before you read an image, `data` is NULL (because the object is empty). So, what you wrote in  https://github.com/3dem/relion/issues/673#issuecomment-701485378 is fine.

> The destructor on the img multidimarray seems to be called at subtractOneParticle() exit, but it looks to have been resized to nothing by then.

This is more relevant. Just before exiting `subtractOneParticle()`, `img` is written to an MRCS stack. Thus `img` has to contain the final subtracted image. Can you set a memory break point to `img().data` and check who destroyed it?

**Comment by bforsbe** [comment]
To me it would seem that the "window" operation here
```
#4 main (argc=9, argv=0x7ffeeba41b28) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/apps/particle_subtract_mpi.cpp:47 (at 0x0000000000514a74)
#3 ParticleSubtractor::run (this=0x7ffeeba3f7f0) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/particle_subtractor.cpp:324 (at 0x00000000005654fc)
#2 ParticleSubtractor::subtractOneParticle (this=0x7ffeeba3f7f0, part_id=4858, imgno=0, counter=2441) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/particle_subtractor.cpp:768 (at 0x000000000056920d)
#1 MultidimArray<double>::window (this=0x7ffeeba3ee40, y0=-128, x0=-128, yF=127, xF=127, init_value=0, n=0) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/multidim_array.h:1452 (at 0x000000000056edbf)
#0 MultidimArray<double>::window (this=0x7ffeeba3ee40, result=..., y0=-128, x0=-128, yF=127, xF=127, init_value=0, n=0) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/multidim_array.h:1435 (at 0x0000000000571563)
```
https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/multidim_array.h#L1449
simply allocates a new array ```result``` without deleting the old one, and sets ``img.data``` to point to the new one.

**Comment by biochem-fan** [comment]
Oh, this is a great discovery of a bug in the core library!

We have to implement `operator =` for MultidimArray.

**Comment by biochem-fan** [comment]
Wait, we do have a copy operator.

https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/multidim_array.h#L4304

`resize()` will call `coreDeallocate()`.

**Comment by bforsbe** [comment]
yes but it doesn't seem like it calls it on the right thing. The resize inside window calls ```coreDeallocate```on ```*this``, which is empty. When window sets ```*this=result```,   ```*this``` is still allocated.

**Comment by bforsbe** [comment]
I'm not sure I was so clear in my last comment. This call 
https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/multidim_array.h#L1172
does nothing in the leaky testcase, becasue in that context there is nothing to deallocate. 

The context is 
```
#6 main (argc=9, argv=0x7ffeeba41b28) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/apps/particle_subtract_mpi.cpp:47 (at 0x0000000000514a74)
#5 ParticleSubtractor::run (this=0x7ffeeba3f7f0) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/particle_subtractor.cpp:324 (at 0x00000000005654fc)
#4 ParticleSubtractor::subtractOneParticle (this=0x7ffeeba3f7f0, part_id=6653, imgno=0, counter=3338) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/particle_subtractor.cpp:768 (at 0x000000000056920d)
#3 MultidimArray<double>::window (this=0x7ffeeba3ee40, y0=-128, x0=-128, yF=127, xF=127, init_value=0, n=0) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/multidim_array.h:1452 (at 0x000000000056edbf)
#2 MultidimArray<double>::window (this=0x7ffeeba3ee40, result=..., y0=-128, x0=-128, yF=127, xF=127, init_value=0, n=0) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/multidim_array.h:1435 (at 0x000000000057158a)
#1 MultidimArray<double>::resize (this=0x7ffeeba3e5c0, Ydim=256, Xdim=256) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/multidim_array.h:1212 (at 0x000000000057ae68)
#0 MultidimArray<double>::resize (this=0x7ffeeba3e5c0, Ndim=1, Zdim=1, Ydim=256, Xdim=256) at /mnt/cephfs/projects/2019032700_Legacy_bjornf/bjornf/relion/relion/src/multidim_array.h:1172 (at 0x000000000051cbb5)
```
When returning to 
https://github.com/3dem/relion/blob/87c294c2783afeb6fd2791449aefcefd3b4ea355/src/multidim_array.h#L1453
```*this``` is still allocated and abandoned for the new pointer.

**Comment by biochem-fan** [comment]
OK, tomorrow I will write a small program that only performs window() and investigate.

**Comment by bforsbe** [comment]
Yes you are right. But the system doesn't see the freed memory. How curious.

**Comment by biochem-fan** [comment]
Isn't it https://stackoverflow.com/questions/52417318/why-does-the-free-function-not-return-memory-to-the-operating-system/52417370?

**Comment by bforsbe** [comment]
Yes but I'm still trying to reconcile this with the **dramatic** difference in performance. Even if I concede that this isn't a leak _per se_ there's still **something** that makes relion run 50 times slower and render nodes incapacitated, that is down to changes in (de)allocation strategy.

**Comment by biochem-fan** [comment]
Can you try compiling with Intel MKLFFT? It provides it's own implementation of `fftw_malloc` (as part of FFTW wrapper). Does the problem persist?

**Comment by bforsbe** [comment]
Will try as soon as possible, probably this weekend.

**Comment by biochem-fan** [comment]
`fftw_malloc` is a wrapper to `memalign`.

https://github.com/FFTW/fftw3/blob/master/kernel/kalloc.c

**Comment by biochem-fan** [comment]
Another case https://www.jiscmail.ac.uk/cgi-bin/wa-jisc.exe?A2=CCPEM;49a32a17.2010.

**Comment by stiobhan** [comment]
I've seen this again recently in our cluster with Relion 3.1.1 (compiled on Ubuntu 18.04 with GCC 8.4, CUDA 10.2, OpenMPI 4.0.4). This uses the FFTW library from the Ubuntu 18.04 repository (v3.3.7-1). I then compiled with Intel MKLFFT and the problem went away, but I will have to test this more thoroughly (possibly also with newer versions of FFTW) to verify that it's not just a coincidence.

**Comment by stiobhan** [comment]
I think I narrowed this down a bit. I can reproduce the issue on our machines (Intel Xeon E5-2690v4) with any FFTW version 3.3.7, 3.3.8 or 3.3.9, when they have been compiled with `--enable-avx` or `--enable-avx2` (like the default FFTW libraries from the Ubuntu repository that are compiled with `--enable-sse2 --enable-avx`). Maybe that's one of the reasons why you couldn't reproduce this before.

On the other hand if I use a library has been compiled with all optimizations enabled (`--enable-sse2 --enable-avx --enable-avx2 --enable-avx-128-fma --enable-avx512`), so that whatever is supported by the CPU gets used at runtime, the job runs fine. A bit odd since on this CPU these extra optimizations cannot be used anyway, but that's the behaviour I see. I use the same RELION binary in each case here and use `LD_LIBRARY_PATH` to switch between the different libraries.

Maybe this helps to reproduce the problem on your side. I wonder if it also depends on the hardware, since FFTW has runtime detection for supported SIMD.

**Comment by stiobhan** [comment]
![image](https://user-images.githubusercontent.com/6123889/106273120-6c4ffb80-6232-11eb-94fb-2b2ae3d2ed7d.png)
here's the behaviour with only 2 MPI processes, where it looks like the node is killing one of the two tasks which releases a load of memory, and the second task happily continues to chew that up until it finally gets killed too.

**Comment by biochem-fan** [comment]
This sounds more and more like an FFTW's issue, not RELION's. At the moment we don't have time to investigate this in detail.

Meanwhile you can use Intel MKLFFT, which is unfortunately not open-source but free of charge and runs faster than FFTW on Intel CPUs.

**Comment by stiobhan** [comment]
> Meanwhile you can use Intel MKLFFT, which is unfortunately not open-source but free of charge and runs faster than FFTW on Intel CPUs.

As I said, I CAN use FFTW if compiled with certain SIMD optimizations.

Anyhow, I think I figured out what's going on here. On each iteration of
https://github.com/3dem/relion/blob/9f3bf18ea82d8532993f1580f479553a200f4ebe/src/particle_subtractor.cpp#L324
around 2280K get allocated on top of the heap. More specifically that happens here
https://github.com/3dem/relion/blob/9f3bf18ea82d8532993f1580f479553a200f4ebe/src/particle_subtractor.cpp#L579
That memory gets released when going out of scope, but at the next iteration some of the free space has already been allocated for other smaller objects. Hence when requesting aligned memory again for `Fsubtract` it does not fit into any gap, so new memory gets allocated to the heap. So at each iteration the heap grows by about 2M (for each MPI process), so the node quickly runs out of memory.

So basically this is a problem of memory fragmentation in the heap memory. By default in `glibc` only allocations of up to 128K should go to the heap, but that is dynamically adjusted:
https://www.gnu.org/software/libc/manual/html_node/Memory-Allocation-Tunables.html#Memory-Allocation-Tunables
> Tunable: glibc.malloc.mmap_threshold
> [...]
> If this tunable is not set, the default value is set to ‘131072’ bytes and the threshold is adjusted dynamically to suit the allocation patterns of the program.

So I suppose memory allocation patterns are different depending on optimizations enabled in the FFTW libraries used, and that is what triggers this behavior here. There is a bug filed for `glibc` to address it, but that won't change anytime soon: https://sourceware.org/bugzilla/show_bug.cgi?id=14581

I'm setting `GLIBC_TUNABLES=glibc.malloc.mmap_threshold=1048576` for now to make sure that everything above 1M gets allocated outside of heap memory, that seems to do the trick.

**Comment by biochem-fan** [comment]
Wow! This is a wonderful analysis! Thank you very much.

We cannot *fix* this from the RELION side; we might be able to reuse `Fsubtract` but other big arrays like `Fimg` might cause troubles next. And such reuse will not work for particles with different box sizes.

I will leave the RELION code as is and suggest your `GLIBC_TUNABLES=glibc.malloc.mmap_threshold=1048576` trick to those unfortunately users who encounter this problem.

**Comment by pur80a** [comment]
Where should I set `GLIBC_TUNABLES=glibc.malloc.mmap_threshold=1048576`? 
During compilation of relion or does it suffice before running relion?

**Comment by biochem-fan** [comment]
This is an environmental variable.
Thus, put the following into your job submission template (assuming your shell is BASH):

```sh
export GLIBC_TUNABLES=glibc.malloc.mmap_threshold=1048576
```

**Comment by zruan** [comment]
I still experience the issue after setting the environmental variable as suggested. I also tried to set `glibc.malloc.mmap_max` to 0, but the issue persists.

**Comment by stiobhan** [comment]
Have you tried smaller values, for example
`GLIBC_TUNABLES=glibc.malloc.mmap_threshold=131072`?

Setting `glibc.malloc.mmap_max` to 0 disables mmap and should make the problem worse as far as I understand. You want to use mmap so that freed space can be returned to the system.

**Comment by zruan** [comment]
Thank you, @stiobhan. Reducing the `glibc.malloc.mmap_threshold` to 2048 seems work on my end.
